added description for operator (#37)
operator-framework/olm-book: docs/what-is-an-operator.md

# What is an Operator?

> Operators are software extensions to Kubernetes that make use of custom resources to manage applications and their components. [Kubernetes documents](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)

Reviewer: ```suggestion
Operators are software extensions to Kubernetes that make use of custom resources to manage applications and their operands. [Kubernetes documents](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/)
```

In the real world, we can define Operators: as a person who operates on some task(machine, equipment, process). Let's say it works on a given task "X". What qualities are required to peform the task X?

We can classify an operator qualities into three categories:

* Observe
* Analyze

> * Act

Reviewer: ```suggestion
In the real world, we can define Operators: as a person who operates on some task(machine, equipment, process). Let's say it works on a given task "X". We can classify an operator qualities into three categories:

* Observe
* Analyze
* Act
```

Observe the X work, analyze the state of X and finally act based on the current state of X to reach the desired state. Which will make the operators task completed, once we attain the desired state.

> Similar work/tasks are performed by an operators. It helps us to extend the Kubernetes API to allow custom functionality or behavior based on the needs of the consumer.

Reviewer: ```suggestion
Similar work/tasks are performed by operators. Operators help us to extend the Kubernetes API to allow custom functionality or behavior based on the needs of the consumer.
```

# Operator Pattern 

> ![Operator internals](images/operator-internals.png)

Reviewer: What tool was used to create this png? Worried about maintainability.

Author: I created the diagram using google-slides. Either I can share the slide link or we can use something else which will be easier to maintain.

* Observe: __Events__
* Analyze: __Current Vs Desired State__
* Act: __Reconciler Loop__

## Components: 

> * Controller: Register the Custom Resource Definition and adds it to the Kubernetes API Schema, exposes a new API end-point for the CRD, added watchers for the namespace to observe, runs a process/loop i.e control loop for acting based on the desired change.

Reviewer: As defined in the [Best practices doc](https://github.com/operator-framework/community-operators/blob/master/docs/best-practices.md#running-on-cluster)- controllers are not supposed to add their CRD to the cluster - could we call the CRD out as a separate component?

> * Controller: Register the Custom Resource Definition and adds it to the Kubernetes API Schema, exposes a new API end-point for the CRD, added watchers for the namespace to observe, runs a process/loop i.e control loop for acting based on the desired change.

Reviewer: This description of a component sounds like a mixture of the CRD Controller and the operator being implemented.

> * Watchers: Allows to observe the events for speific resources type.

Reviewer: Could we include a link to a more concrete description of Kubernetes watchers?

Author: @awgreene: please check the updates and let me know.

* Reconciler: Any modification to the CR is handled by the reconcilation loop. Requeue after conflict/error with an interval defined by the re-sync period. 

> 

Reviewer: Super Nit: please remove this newline.

Author: fixed

## Classification based on deployment type
* Namespace scoped
* Cluster scoped

## Example Operator Workflow

![Operator workflow](images/operator-workflow.png)

> * Defines a namespace(namepace-scoped) for our controller

Reviewer: ```suggestion
* Defines a namespace (namepace-scoped) for our controller
```

> * Creates the CRD: ToDo.

Reviewer: Same as above.

Author: fixed
